### PR TITLE
Comments: Add reply and approve flow

### LINKED ISSUE #15937: Comments: Reply to a pending comment ➡️ the comment is approved.
We need to make sure the comment doesn't disappear, is approved, and the notice should say that it was approved and the reply was successful.

Let's also update the placeholder text to give context to the user.

screen shot 2017-07-06 at 10 17 41 am

### PR DESCRIPTION
**Pull request by Copons**
Closes #15937

Adds the "Reply and Approve" flow to Comment Management.
Replying to an unapproved comment automatically approves it.
A notice appears confirming the comment approval.

jul-07-2017 12-22-49

cc @Automattic/lannister

### CODE REVIEW

File: client/blocks/comment-detail/comment-detail-reply.jsx [hunk 1]
```diff
@@ -35,11 +35,23 @@ export class CommentDetailReply extends Component {
...
-		: 'Reply to comment…';
+	getTextareaPlaceholder = () => {
+		const { authorDisplayName, commentStatus, translate } = this.props;
+
+		if ( 'approved' !== commentStatus ) {
+			return authorDisplayName
+				? translate( 'Approve and reply to %(commentAuthor)s…', {
+					args: { commentAuthor: authorDisplayName }
+				} )
+				: 'Approve and reply to comment…';
```
Comment (dmsnell) [inline client/blocks/comment-detail/comment-detail-reply.jsx]: These should be translated as well…

File: client/blocks/comment-detail/comment-detail-reply.jsx [hunk 2]
```diff
@@ -35,11 +35,23 @@ export class CommentDetailReply extends Component {
...
-		? this.props.translate( 'Reply to %(commentAuthor)s…', {
-			args: { commentAuthor: this.props.authorDisplayName }
-		} )
-		: 'Reply to comment…';
+	getTextareaPlaceholder = () => {
+		const { authorDisplayName, commentStatus, translate } = this.props;
+
+		if ( 'approved' !== commentStatus ) {
+			return authorDisplayName
+				? translate( 'Approve and reply to %(commentAuthor)s…', {
```
Comment (a8ci18n) [inline client/blocks/comment-detail/comment-detail-reply.jsx]: Hi! I've found a possible matching string that has already been translated **2** times:
`translate( 'Reply to %(commentAuthor)s…' )` *ES Score: 7*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/comments/comment-list/index.jsx [hunk 3]
```diff
@@ -181,13 +181,29 @@ export class CommentList extends Component {
...
-		} );
+	submitComment = ( comment, options = { alsoApprove: false } ) => {
+		const { translate } = this.props;
+		const { alsoApprove } = options;
+
+		this.props.removeNotice( `comment-notice-${ comment.parentId }` );
+
+		const noticeMessage = alsoApprove
+			? translate( 'Comment approved and reply submitted.' )
+			: translate( 'Reply submitted.' );
```
Comment (a8ci18n) [inline client/my-sites/comments/comment-list/index.jsx]: Hi! I've found a possible matching string that has already been translated **39** times:
`translate( 'Submitted!' )` *ES Score: 7*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/blocks/comment-detail/comment-detail-reply.jsx [hunk 4]
```diff
@@ -35,11 +35,23 @@ export class CommentDetailReply extends Component {
...
+					args: { commentAuthor: authorDisplayName }
+				} )
+				: translate( 'Approve and reply to comment…' );
+		}
+
+		return authorDisplayName
+			? translate( 'Reply to %(commentAuthor)s…', {
+				args: { commentAuthor: authorDisplayName }
+			} )
+			: translate( 'Reply to comment…' );
```
Comment (a8ci18n) [inline client/blocks/comment-detail/comment-detail-reply.jsx]: Hi! I've found a possible matching string that has already been translated **95** times:
`translate( 'Reply to Comment' )` *ES Score: 11*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/comments/approve-and-reply)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by drw158** [comment]
Does the undo also undo the reply? Or does it just undo the approve action?

**Comment by kwight** [comment]
> Does the undo also undo the reply? Or does it just undo the approve action?

It looks to just undo the parent comment approval (my own reply remained approved after undoing). The Undo messaging implies that the reply would be "unsubmitted" – would that worry users about losing their reply?

**Comment by kwight** [comment]
I wonder if this flow would be clearer and less worrisome if it was more explicit (and more work). Rather than having a reply form for Pending comments, just have a message saying that replies can be made to approved comments. that would lead to an explicit approval, and a separate action of replying. More work for the user, but maybe clearer, and less magical.

**Comment by Copons** [comment]
> It looks to just undo the parent comment approval

Correct.
There's (currently?) no system in place to undo comments.

> I wonder if this flow would be clearer and less worrisome if it was more explicit (and more work). Rather than having a reply form for Pending comments, just have a message saying "Approve this comment to reply" – that would lead to an explicit approval, and a separate action of replying. More work for the user, but maybe clearer, and less magical.

I like this very much!

**Comment by drw158** [comment]
I think I would rather have a "Approve and reply" button than 2 separate actions. But I do like the current solution, which is just updating the placeholder. 

Maybe we could 

- show 2 separate notices for approve and replied. that way, the replied noticed would not have an Undo button
- keep the combined notice, but just get rid of the undo button 

Aside: I wonder if the reply action would be more visible if it was "Reply and approve comment" ?

**Comment by dmsnell** [comment]
> I think that's pretty reasonable for m1. I'm also ok with just clarifying the "undo" text for that combined notice. For example we could replace "undo" there for "unapprove comment".

this I like much better. I don't like undo buttons which not only don't undo the action I did, but also undo the actions I didn't

**Comment by Copons** [comment]
@dmsnell I think I finally got which notice you didn't approve!

Yeah, the flow [`reply + approve` → UNDO → `unapprove but keep reply`] is quite bad, and I didn't realize it.

But about the other one (not related to this PR):
[`like + approve` → UNDO → `unlike + unapprove`], I think it's just correct.

**Pull request merged**